File: Data Structures & Algorithms/permutations/submission-1.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        # ans = []

        # def backtracking(perm):
        #     if not nums:
        #         ans.append(perm[:])
        #         return
            
        #     for i in range(len(nums)):
        #         tmp = nums.pop(i)
        #         perm.append(tmp)
        #         backtracking(perm)
        #         perm.pop()
        #         nums.insert(i,tmp)

        # backtracking([])
        # return ans

        res = []

        def backtacking(comb,valSet):
            if len(comb) == len(nums):
                res.append(comb[:])
                return
            
            for i in range(len(nums)):
                if nums[i] in valSet:
                    continue
                comb.append(nums[i])
                valSet.add(nums[i])
                backtacking(comb,valSet)
                comb.pop()
                valSet.discard(nums[i])
        backtacking([],set())
        return res

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-1.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int[] res = new int[nums.length];
        int[] prefix = new int[nums.length];
        int[] suffix = new int[nums.length];

        int product = 1;
        for(int i = 0;i < nums.length;i++){
            prefix[i] = product;
            product *= nums[i];
        }
        product = 1;
        for(int i = nums.length - 1;i >= 0;i--){
            suffix[i] = product;
            product *= nums[i];
        }

        for(int i = 0;i < nums.length;i++){
            res[i] = prefix[i] * suffix[i];
        }
        return res;
    }
}  
